### ISSUE TITLE
First change event is reported too soon

### ISSUE THREAD

**Issue by MissChocoe** [tinymce/tinymce#6366]
**What is the current behavior? Describe the bug**
On an inline editor, the first caracter you type (or the first modification you make) always trigger the change event.
Susbequents modifications will only trigger the change event on focus out.

**Please provide the steps to reproduce** 
1. Create an inline editor with a listner to the editor "change" event. Put a console.log in there
2. Type one caracter in the editor. The change event is fired.

**What is the expected behavior?**
The first change should not be different from the other and should only trigger the change event on focus out. We've put an ugly work around for now as follow: 

```
                        // firstChange and moreChanges are use to get around ab tinyMCE bug where you get a change event for the first modification (before the loose of focus)
                        // We want to ignore that unless it is the only changed made (moreChanges !== true)
                        var firstChange = true;
                        var moreChanges = false;
                        editor.on('change', function(ed) {
                            if (!firstChange) {
                                // save change to our model
                                moreChanges = true;
                            } else {
                                firstChange = false;
                            }
                        });
                        editor.on('blur', function(ed) {
                            if (!moreChanges && editor.isDirty()) {
                                  // save that only change to our model
                            }
                        });
```

**Which versions of TinyMCE, and which browser / OS are affected by this issue? **
v5.6.2
Same on all tested browser (FF, Chrome and Edge)

**Did this work in previous versions of TinyMCE?**
Unknown

**Comment by mirkootter** [comment]
I just tested some older versions, the error is the same (all 5.x versions and some 4.x versions)

The problem is even more severe:
* Start with an empty editor
* Type some letters inside (one is enough). An ```onChange``` event is fired
* Remove those letters again
* Focus something else. **No onChange is fired now**

The same happens if the editor is not empty by default: The final onChange does not fire if the original input is restored at the end.

In short: The user did not change anything, but there was exactly one onChange event **with** a change, so my application is now in inconsistent state...

**Comment by mirkootter** [comment]
I did some debugging and found this (modules/tinymce/src/core/main/ts/undo/Setup.ts#L73)
```typescript
// Fire a TypingUndo/Change event on the first character entered
if (isFirstTypedCharacter.get() && undoManager.typing && Levels.isEq(Levels.createFromEditor(editor), undoManager.data[0]) === false) {
  if (editor.isDirty() === false) {
    editor.setDirty(true);
    editor.fire('change', { level: undoManager.data[0], lastLevel: null });
  }

  editor.fire('TypingUndo');
  isFirstTypedCharacter.set(false);
  editor.nodeChanged();
}
```
Due to the comment it seems that this behaviour is intended. However, the code also shows a simple workaround:
```javascript
function onChange(ev) {
  if(!ev.lastLevel) return; // early out on first change

  console.log("this change is what you expect");
}
```

**Comment by MissChocoe** [comment]
Thanks @mirkootter . This workaround is a little bit better than what I did but I still need to save that first caracter if it is the only one. Not sure why they need that, but they could probably set the editor dirty without firing the change event.

**Comment by metricjs** [comment]
Hi @MissChocoe @mirkootter, thanks for reporting this. We are still discussing this ticket internally, but to help us decide what should happen with this event, we wondered what your use case is? Why are you listening to the change event, and why is the first fire of it causing problems for you?

**Comment by mirkootter** [comment]
Hi,
my use case is simple: I want to use the ```onChange``` to detect changes, i.e. I want to know when the user changed the content of the editor, for example to update my data model. **My user does not have an OK button or anything to press if he has finished**.

It is completely fine that the change event does not fire on every key press, that is even a benefit. But it should fire on every change.

In fact, even the first ```onChange``` is not the problem. The problem is that the final one is missing.
- The editor contains the text "Hello!" to begin with
- The user adds a second '!' to it.
- Now, the user decides otherwise and removes the '!' again.
- Now, he is satisfied and leaves the editor (**without a change**)

Of course, the user now expects that nothing was changed.

However, my internal data model is now "Hello!!", because there was exactly one ```onChange``` event: the first one.

**Comment by MissChocoe** [comment]
My use case is similar. We save the content of the editor in our internal model and we don't have any 'save' button. 

The reason why the first change with only a single entry causes us probleme is that we use that initial value to initialzed another field too. So I need it to be updated with something more than a single caracter.

@mirkootter For your specific issue, we listen to undo/redo event from tinymce to save the content in those cases

**Comment by metricjs** [comment]
@mirkootter @MissChocoe Thanks for clarifying your use cases! I'll discuss this with the team, and get back to you when we make a decision about what we want to do.

**Comment by metricjs** [comment]
@MissChocoe After discussing your case, the team agrees that that first `change` event should not fire. I'll log a ticket internally for us to change that. However I did want to let you know that since we follow [Semantic Versioning](https://semver.org/) we may not be able to change it until 6.0, as it may be a breaking change for other customers. We will let you know when it is scheduled for a release.

@mirkootter It sounds like your problem is more that there is no `change` event fired on focus out? From what we understand of your use case it should be, so that sounds like a bug. See [this fiddle](https://fiddle.tiny.cloud/c3haab/2) for an example of how this works (open the console, type into the editor then click elsewhere on the page - you should see the relevant events being fired). If you are seeing this problem, could you please make a new issue (so we can track the two separately) and please include the following details if possible:

- A fiddle that demos how to replicate your problem. You can use our [Fiddle](https://fiddle.tiny.cloud/) site, or Codepen or any similar site. 
- Details about any integrations you're using (e.g. React).
- What happens when your users "leave the editor" (e.g. Are they clicking a button to navigate elsewhere? Are they closing a popup?)?
- You said `Now, the user decides otherwise and removes the '!' again.` - to clarify, you mean deleting or backspacing characters, not undoing them?
- Anything else that might help us replicate your issue.

I've also included in the [fiddle I linked above](https://fiddle.tiny.cloud/c3haab/2) a demo of the `beforeunload` event. From what you have said, we thought it may be helpful for your use case. Unfortunately, we're unsure what would be best based on the information we currently have though.

**Comment by cdev003** [comment]
sadsad
